### PR TITLE
Some broker pods get stuck caused by a terminated container when there is no healthy broker in the kafka cluster.

### PR DESCRIPTION
**Pull request by bartam1**
| Q               | A      |
| --------------- | ------ |
| Bug fix?        | yes |
| New feature?    | no |
| API breaks?     | no |
| Deprecations?   | no |
| License         | Apache 2.0 |

### What's in this PR?
 If reconcilePerBrokerDynamicConfig gets an error then I let the loop continue to the next broker. A combined error will be returned after the broker loop so the operator will try again later. 

### Why?
When in a kafka cluster there is not any working kafka broker present, but there are multiple broker pods with terminated container, only one broker pod will be restarted. Other brokers with terminated containers will not until  at least one healthy broker will be present in the kafka cluster. The root cause of this that brokers are unreachable and reconcilePerBrokerDynamicConfig function can't set dynamic configs and It will be ended with an error.

### Checklist
- [x] Implementation tested
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/developer-guide/blob/master/docs/coding-style/error-handling-guide.md)
- [x] Logging code meets the guideline

### CODE REVIEW

**Review by Kuvesz — APPROVED** [review]
LGTM

File: pkg/resources/kafka/kafka.go [hunk 1]
```diff
@@ -283,11 +284,18 @@ func (r *Reconciler) Reconcile(log logr.Logger) error {
 		if err = r.updateStatusWithDockerImageAndVersion(broker.Id, brokerConfig, log); err != nil {
 			return err
 		}
+		// If dynamic configuration gets an error then let the loop continue to the next broker
+		// combined error will be returned after the broker loop. This solve that case when other brokers could get healthy,
+		// but the loop exits too soon because brokers are unreachable and dynamic configs cannot be set.
 		if err = r.reconcilePerBrokerDynamicConfig(broker.Id, brokerConfig, configMap, log); err != nil {
-			return err
+			perBrokerDynamicConfigCombineError = errors.Combine(perBrokerDynamicConfigCombineError, err)
```
Comment (hi-im-aren) [inline pkg/resources/kafka/kafka.go:291]: We could add some details for the error like `broker.Id, brokerConfig, configMap`

**Review by pregnor — DISMISSED** [review]
LGTM

**Review by Kuvesz — DISMISSED** [review]

File: pkg/resources/kafka/kafka.go [hunk 2]
```diff
@@ -283,11 +284,21 @@ func (r *Reconciler) Reconcile(log logr.Logger) error {
...
+		err = r.reconcilePerBrokerDynamicConfig(broker.Id, brokerConfig, configMap, log)
+		if err != nil {
+			log.Error(err, "setting dynamic configs has failed", "brokerID", broker.Id)
+			isPerBrokerDynamicConfigFailed = true
 		}
 	}
 
+	if isPerBrokerDynamicConfigFailed {
+		// re-reconcile to retry setting the dynamic configs
+		return errors.New("setting dynamic configs for some brokers has failed.")
```
Comment (stoader) [inline pkg/resources/kafka/kafka.go:299]: Add context to the error (e.g. see https://github.com/banzaicloud/koperator/blob/master/pkg/resources/kafka/kafka.go#L304-L307)

File: pkg/resources/kafka/kafka.go [hunk 3]
```diff
@@ -240,6 +240,7 @@ func (r *Reconciler) Reconcile(log logr.Logger) error {
 	}
 
 	reorderedBrokers := reorderBrokers(brokerPods, r.KafkaCluster.Spec.Brokers, r.KafkaCluster.Status.BrokersState, controllerID)
+	isPerBrokerDynamicConfigFailed := false
```
Comment (stoader) [inline pkg/resources/kafka/kafka.go:243]: How about `allBrokerDynamicConfigSucceeded := true` and than if an error occurs set it to `false`

**Review by stoader — CHANGES_REQUESTED** [review]

**Review by pregnor — DISMISSED** [review]

**Review by stoader — DISMISSED** [review]

**Review by pregnor — APPROVED** [review]

**Comment by bartam1** [comment]
I have disabled the cyclo check for the reconcile function temporarily.
Im going to make refactor for the reconcile function and put back cyclo check in another PR

**Review by stoader — APPROVED** [review]

**Pull request merged**